### DISCUSSION TITLE
allow to add `--gpus all` and add Local AI community container with GPU support

### DISCUSSION THREAD

**Discussion by luzfcb** [nextcloud/all-in-one#4277] [Ideas] [answered]
Questions:

1 - How to force the local-ai community container to use a nvidia GPU-enabled tag like: `v2.9.0-cublas-cuda12-ffmpeg` instead of the regular `v2.0.0-ffmpeg-core` tag?

2 - For the Nvidia GPU to be able to be used by a container based on the image-based `quay.io/go-skynet/local-ai:v2.9.0-cublas-cuda12-ffmpeg` it's required to use the `--gpus all` on the `docker run` command

How to configure the Nextcloud AIO to include the `--gpus all` argument on the docker run command of Nextcloud AIO Local-Ai community container?

https://github.com/mudler/LocalAI/blob/master/docs/content/docs/features/GPU-acceleration.md
https://docs.docker.com/config/containers/resource_constraints/#gpu

Note: v2.9.0-cublas-cuda12-ffmpeg tag has more than 16GB in size

**Comment by szaimen** [comment]
Hi, in order to make this work, first a capability for AIO would need to be added to add all gpus to a container. See https://github.com/nextcloud/all-in-one/pull/1659 as inspiration. 

When that is done, one (you?) could add an additional community container, e.g. name it local-ai-gpu or something that itself is built from `v2.9.0-cublas-cuda12-ffmpeg`. See https://github.com/nextcloud/all-in-one/pull/3495 and https://github.com/szaimen/aio-local-ai as inspiration

**Comment by szaimen** [comment, marked as answer]
Nvidia gpu passthrough is now being worked on in https://github.com/nextcloud/all-in-one/pull/5132

**Reply by szaimen** [reply]
This is now released with v10.2.0 Beta. Testing and feedback is welcome! See https://github.com/nextcloud/all-in-one#how-to-switch-the-channel

**Comment by gbertolottiKS** [comment]
Hi @szaimen ,
I see the work needed to use the GPU was merged, thanks to all the contributors! 
So now I have few questions:
1) what GPUs are supported ? Only NVDIA ones?
2) is it necessary to activate something to be able to use the GPUs or it's on by default?
3) once Docker and Nextcloud sees the GPU, what are the steps to follow to use the local AI? I know this guide (https://github.com/nextcloud/all-in-one/tree/main/community-containers/local-ai): is it the only source of truth or we need to follow also some other steps?
4) in a previous attempt to use the AI capabilities, we see that it exist Local AI AND some external apps (ExApp): what are the difference between the two? Are they alternatives or are complementary? We strive to find a guide that explains this

Thanks

**Reply by szaimen** [reply]
> * what GPUs are supported ? Only NVDIA ones?

In general all of them by using https://github.com/nextcloud/all-in-one?tab=readme-ov-file#with-open-source-drivers-mesa-for-amd-intel-and-new-drivers-nouveau-for-nvidia

> * is it necessary to activate something to be able to use the GPUs or it's on by default?

Yes, you need to configure two variables: https://github.com/nextcloud/all-in-one?tab=readme-ov-file#with-proprietary-drivers-for-nvidia-warning-beta and https://github.com/nextcloud/all-in-one?tab=readme-ov-file#with-open-source-drivers-mesa-for-amd-intel-and-new-drivers-nouveau-for-nvidia

> * once Docker and Nextcloud sees the GPU, what are the steps to follow to use the local AI? I know this guide (https://github.com/nextcloud/all-in-one/tree/main/community-containers/local-ai): is it the only source of truth or we need to follow also some other steps?

This community container does not support gpus yet. See https://github.com/nextcloud/all-in-one/pull/5795 and https://github.com/nextcloud/all-in-one/pull/5797 for WIPs that add a different local-ai image that allows this.

> * in a previous attempt to use the AI capabilities, we see that it exist Local AI AND some external apps (ExApp): what are the difference between the two? Are they alternatives or are complementary? We strive to find a guide that explains this

external apps ship different AI features than Local AI. If you want all AI features, you need to enable both.
